codeworld-requirements/src/Framework.hs

```diff
@@ -0,0 +1,129 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE PackageImports #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TupleSections #-}
+
+{-
+  Copyright 2019 The CodeWorld Authors. All rights reserved.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+-}
+
+module Framework where
+
+import Control.Applicative
+import Control.Concurrent
+import Control.Monad
+import Control.Monad.Catch
+import Control.Monad.IO.Class
+import Control.Monad.State
+import Data.ByteString (ByteString)
+import qualified Data.ByteString as B
+import Data.List (isPrefixOf, intercalate, foldl')
+import Data.Map (Map)
+import qualified Data.Map as M
+import Data.Maybe
+import Data.Monoid
+import Data.Text (Text)
+import qualified Data.Text as T
+import Data.Text.Encoding (decodeUtf8, encodeUtf8)
+import Language.Haskell.Exts
+import System.Exit (ExitCode(..))
+import System.Directory
+import System.FilePath
+import System.IO
+import System.IO.Temp (withSystemTempDirectory)
+import System.Process
+
+import qualified Config as GHCParse
+import qualified DynFlags as GHCParse
+import qualified FastString as GHCParse
+import qualified Fingerprint as GHCParse
+import qualified GHC.LanguageExtensions.Type as GHCParse
+import qualified HeaderInfo as GHCParse
+import qualified HsExtension as GHCParse
+import qualified HsSyn as GHCParse
+import qualified HscTypes as GHCParse
+import qualified Lexer as GHCParse
+import qualified Panic as GHCParse
+import qualified Parser as GHCParse
+import qualified Platform as GHCParse
+import qualified SrcLoc as GHCParse
+import qualified StringBuffer as GHCParse
+
+data GHCParsedCode = GHCParsed (GHCParse.HsModule GHCParse.GhcPs) | GHCNoParse
+
+ghcExtensionsByName :: Map String GHCParse.Extension
+ghcExtensionsByName = M.fromList [
+    (GHCParse.flagSpecName spec, GHCParse.flagSpecFlag spec)
+    | spec <- GHCParse.xFlags ]
+
+applyExtensionToFlags :: GHCParse.DynFlags -> String -> GHCParse.DynFlags
+applyExtensionToFlags dflags name
+  | "No" `isPrefixOf` name =
+        GHCParse.xopt_unset dflags $ fromJust $ M.lookup (drop 2 name) ghcExtensionsByName
+  | otherwise =
+        GHCParse.xopt_set dflags $ fromJust $ M.lookup name ghcExtensionsByName
+
+ghcParseCode :: [String] -> Text -> GHCParsedCode
+ghcParseCode exts src = do
+    let buffer = GHCParse.stringToStringBuffer (T.unpack src)
+        defaultFlags = GHCParse.defaultDynFlags fakeSettings fakeLlvmConfig
```

Comment:
I think you want to start with the DynFlags that were passed in, right?  It doesn't seem right to use a fake config when there's a real one.

Reply:
Yes, I think you're right on this. The only place this is currently used is by the template matcher for parsing the templates, so no situation came up in testing where I noticed it was still using the fake ones! But it'll probably end up getting used for a couple more things where it will matter, so I'll fix it, although it'll involve passing the DynFlags around to a few functions.

Reply:
Seconded (thirded?). A lot of subtle bugs can arise from using fake `DynFlags`.

Reply:
Thinking about this more... there's some subtlety here.  The global DynFlags contains things from LANGUAGE and OPTIONS_GHC pragmas in the current module, right?  That means it's possible for student source code to change the interpretation of the requirements, and possibly cause them to fail to be parsed?

This isn't a new problem, though.  If you agree this is an issue, maybe create an issue to track it, but don't block this pull request on fixing it.  It would even be fine if it didn't get fixed this summer.  I just want to keep track of the known limitation.

Reply:
You could just override those, no? Wouldnt setting those to nothing be preferred?

Reply:
I don't think just categorically overriding them would work - theoretically we want this to be usable for code that makes use of language extensions etc, and it'd be more difficult to tell the difference between ones that are being used for good reasons and ones that will break the requirements. Maybe it could be a temporary solution, though! I'll make an issue as suggested.